rust/worker/src/sysdb/sysdb.rs

```diff
@@ -0,0 +1,237 @@
+        collection: Option<Uuid>,
+    ) -> Result<Vec<Segment>, GetSegmentsError>;
+}
+
+// We'd like to be able to clone the trait object, so we need to use the
+// "clone box" pattern. See https://stackoverflow.com/questions/30353462/how-to-clone-a-struct-storing-a-boxed-trait-object#comment48814207_30353928
+// https://chat.openai.com/share/b3eae92f-0b80-446f-b79d-6287762a2420
+pub(crate) trait SysDbClone {
+    fn clone_box(&self) -> Box<dyn SysDb>;
+}
+
+impl<T> SysDbClone for T
+where
+    T: 'static + SysDb + Clone,
+{
+    fn clone_box(&self) -> Box<dyn SysDb> {
+        Box::new(self.clone())
+    }
+}
+
+impl Clone for Box<dyn SysDb> {
+    fn clone(&self) -> Box<dyn SysDb> {
+        self.clone_box()
+    }
+}
+
+#[derive(Clone)]
+// Since this uses tonic transport channel, cloning is cheap. Each client only supports
+// one inflight request at a time, so we need to clone the client for each requester.
+pub(crate) struct GrpcSysDb {
+    client: sys_db_client::SysDbClient<tonic::transport::Channel>,
+}
+
+impl GrpcSysDb {
+    pub(crate) async fn new() -> Self {
+        let client = sys_db_client::SysDbClient::connect("http://[::1]:50051").await;
+        match client {
+            Ok(client) => {
+                return GrpcSysDb { client: client };
+            }
+            Err(e) => {
+                // TODO: config error
+                panic!("Failed to connect to sysdb: {}", e);
+            }
+        }
+    }
+}
+
+#[async_trait]
+impl SysDb for GrpcSysDb {
+    async fn get_collections(
+        &mut self,
+        collection_id: Option<Uuid>,
+        topic: Option<String>,
+        name: Option<String>,
+        tenant: Option<String>,
+        database: Option<String>,
+    ) -> Result<Vec<Collection>, GetCollectionsError> {
+        // TODO: move off of status into our own error type
+        let collection_id_str;
+        match collection_id {
+            Some(id) => {
+                collection_id_str = Some(id.to_string());
+            }
+            None => {
+                collection_id_str = None;
+            }
+        }
+
+        let res = self
+            .client
+            .get_collections(chroma_proto::GetCollectionsRequest {
+                id: collection_id_str,
+                topic: topic,
+                name: name,
+                tenant: if tenant.is_some() {
+                    tenant.unwrap()
+                } else {
+                    DEFAULT_TENANT.to_string()
+                },
+                database: if database.is_some() {
+                    database.unwrap()
+                } else {
+                    DEFAULT_DATBASE.to_string()
+                },
+            })
+            .await;
+
+        match res {
+            Ok(res) => {
+                let collections = res.into_inner().collections;
+
+                let collections = collections
+                    .into_iter()
+                    .map(|proto_collection| proto_collection.try_into())
```

Comment:
.try_into is satisfied by the impls of Try_From(). Cool!

Reply:
🦀